Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/DocumentAdapter.h

```diff
@@ -22,6 +22,75 @@ namespace AZ::DocumentPropertyEditor
     using DocumentAdapterPtr = AZStd::shared_ptr<DocumentAdapter>;
     using ConstDocumentAdapterPtr = AZStd::shared_ptr<const DocumentAdapter>;
 
+    //! A message, invoked via CallbackAttribute, that is routed to a DocumentAdapter.
+    //! Adapters, or their views, may handle these messages as they see fit, possibly
+    //! providing a response.
+    struct AdapterMessage
+    {
+        //! The name of this message (derived from the CallbackAttribute's name)
+        AZ::Name m_messageName;
+        //! The path in the adapter from which this message originated
+        Dom::Path m_messageOrigin;
+        //! A DOM array value containing marshalled arguments for this message.
+        Dom::Value m_messageParameters;
+        //! An arbitrary, user-specified DOM value containing additional context data.
+        //! This is used to supplement message parameters on the adapter level.
+        Dom::Value m_contextData;
+
+        bool Match([[maybe_unused]] Dom::Value& result) const
+        {
+            return false;
+        }
+
+        template <typename CallbackAttribute, typename Callback, typename... Rest>
+        bool Match(Dom::Value& result, const CallbackAttribute& attribute, const Callback& callback, Rest... rest) const
+        {
+            if (attribute.MatchMessage(*this, result, callback))
+            {
+                return true;
+            }
+            return Match(result, rest...);
+        }
+
+        //! Takes arguments in pairs of {attribute, callback} and invokes the associated callback
+        //! if this message is applicable.
+        //! \return The value returned by the message handler, or null if no messages matched.
+        template <typename CallbackAttribute, typename Callback, typename... Rest>
+        Dom::Value Match(const CallbackAttribute& attribute, const Callback& callback, Rest... rest) const
+        {
+            Dom::Value result;
+            Match(result, attribute, callback, rest...);
+            return result;
+        }
+    };
+
+    //! An adapter message bound to a given adapter, to be invoked as part of a CallbackAttribute.
+    //! Used to store a message and all of the associated context as part of a DocumentAdapter.
+    //! \see AdapterBuilder::AddMessageHandler
+    struct BoundAdapterMessage
+    {
+        DocumentAdapter* m_adapter;
+        AZ::Name m_messageName;
+        Dom::Path m_messageOrigin;
+        Dom::Value m_contextData;
+
```

Comment:
Since this is a public structure, the DocumentAdapter should be initialized to nullptr